cowprotocol/cowswap
# Add library and UI library

# Summary


This PRs makes a bunch of changes to make CoW Swap be  a `mono-repo`

To test how it works, I created 2 projects:
* `ui`: UI library, meant to be reused among projects
* `uI-utils`: handy libs to be used in other projects

## Dependencies
Now that we have more than one project in this git repo, it would be nice to visualise the dependencies between the projects.

Fortunately, NX makes this chart for us just based on the imports
<img width="512" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/07553126-cf11-416f-a4cd-991e15e1792e">

Once we set our rules on which projects can depend on what, we will be able to enforce our architecture better!

## UI Library + UI utils
To showcase how to do a re-usable library of components, i created a dummy component that just adds a pink title. This component will use also a util. In summary:
* CoW Swap now has a page `/kitchen-sink` which uses the `<PinkTitle />` component
* This component is not in `CoW Swap`, it is in an independent project, part of the monorepo, called `ui` (`@cowprotocol/ui`)
* The component itself uses a function `return5` exported by a different project, also part of the monorepo` called `ui-utils`. 

You can use this locally, and it will work out of the box, with live reloading.

<img width="1893" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/e1fa755d-43cd-45a0-ac9f-3b670d8464cb">

## How I see this evolving 🔮

I would like to kill the `CoW Swap cra app`, the idea would be:
- Start to create small libraries, and move code there
- Once CoW Swap is minimal, we could create a next.js (for example) app and inject the components in our libs
- Hopefully by then, we can kill the original CoW Swap, and the new app will be way easier to maintain and will have WAY FASTER ci build, and also will be able to share code with other apps


## Struggles during this PR

This is some things I struggled with, I kept bumping myself against the wall

- [x] tsconfig was conflicting between cowswap and the new projects
- [x] importing modules from NX wouldn't work out of the box, this is because CRA won't use the resolution from `tsconfig`, so we need to instruct webpack how to resolve them (by using aliases in CRACO config)
- [x] We can't use for now a more `monorepo` style organization, all our subprojects needs to be in `src` if we want them to work with CRA! I tried many things, like overriding with CRACO, tsconfig, and JEST config, but ultimately it gives you an error you can't get rid off!
- [x] eslint of our project is incompatible with NX one, once we get rid of CRACO/CRA we will be able to use only nx
- [x] Circular dependency in the 
- [x] Because I added the projects in the `src` of CoW Swap, now the test run by default. CRA also don't let me add an ignore config param (its one of the forbidden configs). I solved by adding an argument to the test to ignore `src/libs`, I had to declare it in `project.json` so now i can run it with `nx test cowswap`
- [x] The tests were not working for both cowswap and for the subprojects. The configs were conflicting
- [x] E2e where not working

## Running all tests
Finally, they work for the "legacy CRA" and for the new libs!

<img width="653" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/a629c0de-820d-4df8-b23d-084420450a5f">




# To Test
- You can check the new component being used here: https://swap-dev-git-add-library-cowswap.vercel.app/#/kitchen-sink?chain=mainnet 
- You can test this locally and try to see if live reloading works
      - https://localhost:3000/#/kitchen-sink?chain=mainnet 
- You can try to run the tests and build for `ui`: `npx nx build ui` or `npx nx test ui`
- You can try to run the tests and build for `ui-utils`: `npx nx build ui-utils` or `npx nx test ui-utils`
- Run all tests! `yarn test`
- Run all builds: `yarn build`
